---
title: "Investigating Pertussis Resurgence"
author: "Loryn Young"
date: "3/9/2022"
output: pdf_document
---

1. Investigating pertussis cases by year

```{r}
cdc <- data.frame(
                                 Year = c(1922L,1923L,1924L,1925L,
                                          1926L,1927L,1928L,1929L,1930L,1931L,
                                          1932L,1933L,1934L,1935L,1936L,
                                          1937L,1938L,1939L,1940L,1941L,1942L,
                                          1943L,1944L,1945L,1946L,1947L,
                                          1948L,1949L,1950L,1951L,1952L,
                                          1953L,1954L,1955L,1956L,1957L,1958L,
                                          1959L,1960L,1961L,1962L,1963L,
                                          1964L,1965L,1966L,1967L,1968L,1969L,
                                          1970L,1971L,1972L,1973L,1974L,
                                          1975L,1976L,1977L,1978L,1979L,1980L,
                                          1981L,1982L,1983L,1984L,1985L,
                                          1986L,1987L,1988L,1989L,1990L,
                                          1991L,1992L,1993L,1994L,1995L,1996L,
                                          1997L,1998L,1999L,2000L,2001L,
                                          2002L,2003L,2004L,2005L,2006L,2007L,
                                          2008L,2009L,2010L,2011L,2012L,
                                          2013L,2014L,2015L,2016L,2017L,2018L,
                                          2019L),
         No..Reported.Pertussis.Cases = c(107473,164191,165418,152003,
                                          202210,181411,161799,197371,
                                          166914,172559,215343,179135,265269,
                                          180518,147237,214652,227319,103188,
                                          183866,222202,191383,191890,109873,
                                          133792,109860,156517,74715,69479,
                                          120718,68687,45030,37129,60886,
                                          62786,31732,28295,32148,40005,
                                          14809,11468,17749,17135,13005,6799,
                                          7717,9718,4810,3285,4249,3036,
                                          3287,1759,2402,1738,1010,2177,2063,
                                          1623,1730,1248,1895,2463,2276,
                                          3589,4195,2823,3450,4157,4570,
                                          2719,4083,6586,4617,5137,7796,6564,
                                          7405,7298,7867,7580,9771,11647,
                                          25827,25616,15632,10454,13278,
                                          16858,27550,18719,48277,28639,32971,
                                          20762,17972,18975,15609,18617)
       )
cdc
```

```{r}
library(ggplot2)
```

> Q1. With the help of the R “addin” package datapasta assign the CDC pertussis case number data to a data frame called cdc and use ggplot to make a plot of cases numbers over time.

```{r}
ggplot(cdc) +
  aes(x=Year, y=No..Reported.Pertussis.Cases) +
  geom_point() +
  geom_line() +
  labs(title="Pertussis Cases by Year (1922-2019",
       x="Year", 
       y="Number of cases")
```

2. A tale of two vaccines (wP & aP)

> Q2. Using the ggplot geom_vline() function add lines to your previous plot for the 1946 introduction of the wP vaccine and the 1996 switch to aP vaccine (see example in the hint below). What do you notice?

After adding lines to the plot, I notice that the cases significantly decreased after the 1946 introduction of the wP vaccine. After the introduction of the 1996 aP vaccine, they subtly increased, but still stayed much lower than they were before the vaccines were introduced. 

```{r}
ggplot(cdc) +
  aes(x=Year, y=No..Reported.Pertussis.Cases) +
  geom_point() +
  geom_line() +
  labs(title="Pertussis Cases by Year (1922-2019",
       x="Year", 
       y="Number of cases") +
  geom_vline(xintercept=c(1946, 1996), linetype='dashed', color=c('blue', 'red'))
```
> Q3. Describe what happened after the introduction of the aP vaccine? Do you have a possible explanation for the observed trend?

After the introduction of the aP vaccine, the number of cases stayed much lower than the number of cases before a vaccine was introduced. However, the number of cases did increase compared to the plateu before the vaccine was introduced. This may be a sign that the aP vaccine was not as effective as the wP vaccine, or that there was an outbreak in the early 2000s. 

3. Exploring CMI-PB data

```{r}
library(jsonlite)
```

```{r}
subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = TRUE) 
subject
```

> Q4. How may aP and wP infancy vaccinated subjects are in the dataset?

47 aP infancy vaccinated subjects and 49 wP infancy vaccinated subjects 

```{r}
head(subject, 3)
table(subject$infancy_vac)
```

> Q5. How many Male and Female subjects/patients are in the dataset?

There are 66 females and 30 males

```{r}
table(subject$biological_sex)
```


> Q6. What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc…)?

```{r}
table(subject$biological_sex, subject$race)
```

```{r}
library(lubridate)
```

What is today’s date (at the time I am writing this obviously)

```{r}
today()
```

How many days have passed since new year 2000?

```{r}
today() - ymd("2000-01-01")
```

What is this in years?

```{r}
time_length( today() - ymd("2000-01-01"),  "years")
```

> Q7. Using this approach determine (i) the average age of wP individuals, (ii) the average age of aP individuals; and (iii) are they significantly different?

(i) Average age of wP individuals: 35
(ii) Average age of aP individuals: 24
(iii) Yes, they are significantly different (>10 yrs apart)

```{r}
subject$age <- today() - ymd(subject$year_of_birth)
subject$age
```

```{r}
library(dplyr)
```
```{r}
ap <- subject %>% filter(infancy_vac == "aP")
round( summary( time_length( ap$age, "years" ) ) )
```

```{r}
wp <- subject %>% filter(infancy_vac == "wP")
round( summary( time_length( wp$age, "years" ) ) )
```

> Q8. Determine the age of all individuals at time of boost?

```{r}
int <- ymd(subject$date_of_boost) - ymd(subject$year_of_birth)
age_at_boost <- time_length(int, "year")
head(age_at_boost)
```

> Q9. With the help of a faceted boxplot (see below), do you think these two groups are significantly different?

Yes, I do think these two groups are significantly different. The boxplot shows that the aP vaccine is typically given at a younger age than the wP vaccine. Additionally, the p-value is less than 0.05, indicating the two groups are signficantly different.

```{r}
ggplot(subject) +
  aes(time_length(age, "year"),
      fill=as.factor(infancy_vac)) +
  geom_histogram(show.legend=FALSE) +
  facet_wrap(vars(infancy_vac), nrow=2) 
```

```{r}
x <- t.test(time_length( wp$age, "years" ),
       time_length( ap$age, "years" ))

x$p.value
```

```{r}
specimen <- read_json("https://www.cmi-pb.org/api/specimen", simplifyVector = TRUE) 
titer <- read_json("https://www.cmi-pb.org/api/ab_titer", simplifyVector = TRUE)
```

```{r}
specimen
```
```{r}
titer
```

> Q9. Complete the code to join specimen and subject tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```
> Q10. Now using the same procedure join meta with titer data so we can further analyze this data in terms of time of visit aP/wP, male/female etc.

```{r}
abdata <- inner_join(titer, meta)
dim(abdata)
head(abdata)
```
> Q11. How many specimens (i.e. entries in abdata) do we have for each isotype?

```{r}
table(abdata$isotype)
```

> Q12. What do you notice about the number of visit 8 specimens compared to other visits?

The number of visit 8 specimens compared to other visits is much lower. 

```{r}
table(abdata$visit)
```

4. Examine IgG1 Ab titer levels

```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```

> Q13. Complete the following code to make a summary boxplot of Ab titer levels for all antigens:

```{r}
ggplot(ig1) +
  aes(ab_titer, antigen) +
  geom_boxplot() + 
  facet_wrap(vars(visit), nrow=2)
```

> Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?

FIM2/3 shows a large difference in the level of IgG1 antibody titers recognizing them over time. This may be because this antigen has more binding sites for the antibody titers than other antigens, ultimately resulting in more recognition. 

```{r}
ggplot(ig1) +
  aes(ab_titer, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(visit), nrow=2) +
  theme_bw()
```

```{r}
ggplot(ig1) +
  aes(ab_titer, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(infancy_vac, visit), nrow=2)
```

> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment).

```{r}
filter(ig1, antigen=="FHA") %>%
  ggplot() +
  aes(ab_titer, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title="FHA antigen levels per visit (aP red, wP teal)")
```

```{r}
filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(ab_titer, col=infancy_vac) +
  geom_boxplot(show.legend = FALSE) +
  facet_wrap(vars(visit)) +
  theme_bw() +
  labs(title="FIM2/3 antigen levels per visit (aP red, wP teal)")
```

> Q16. What do you notice about these two antigens time course and the FIM2/3 data in particular?

FHA levels stay fairly consistent over time, whereas FIM2/3 heavily rise over time and exceed FHA levels. Both FHA and FIM2/3 levels peak at visit 5 and then decline. 

> Q17. Do you see any clear difference in aP vs. wP responses?

There is no clear difference in aP and wP responses in FHA or in the FIM2/3 data. The wP and aP subjects tend to have a similar trend, slightly increasing over time and slowly declining after visit 5.

5. Obtaining CMI-PB RNASeq data

```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"

rna <- read_json(url, simplifyVector = TRUE) 
```

```{r}
ssrna <- inner_join(rna, meta)
```

> Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm).

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

> Q19. What do you notice about the expression of this gene (i.e. when is it at it’s maximum level)?

The expression of this gene rapidly increases right before reaching its maximum level and rapidly decreases immediately after reaching its maximum level. Maximum level of expression of this gene is during visit 4. Outside of visit 4, the expression is consistently very low. 

> Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?

No, this pattern does not match the trend of antibody titer data because cells make antibodies, which are long lived. This means there is not a rapid decline after reaching the peak, whereas in this plot, there is a clear decrease in tpm just one visit after reaching the peak. 

```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit))
```

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug()
```

6. Working with larger datasets 

```{r}
rnaseq <- read.csv("~/Downloads/2020LD_rnaseq.csv")

head(rnaseq,3)
```
```{r}
dim(rnaseq)
```

```{r}
n_genes <- table(rnaseq$specimen_id)
head( n_genes , 10)
```

How many specimens?

```{r}
length(n_genes)
```

Check if there are the same number of genes for each specimen

```{r}
all(n_genes[1]==n_genes)
```

Convert to "wide" format

```{r}
library(tidyr)

rna_wide <- rnaseq %>% 
  select(versioned_ensembl_gene_id, specimen_id, tpm) %>%
  pivot_wider(names_from = specimen_id, values_from=tpm)

dim(rna_wide)
```

```{r}
head(rna_wide[,1:7], 3)
```

