---
title: "Class 9: Structural Bioinformatics"
author: "Loryn Young"
date: "2/15/2022"
output: pdf_document
---

#View(Data.Export.Summary)

> Q1. What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

X-ray: 87.1% (163154/187170)
Electron Microscopy: 5.3% (10086/187170)

> Q2. What proportion of structures in the PDB are protein?

(163330/187170) = 87.3% of structures are protein 

> Q3. Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

It is not easy to determine how many HIV-1 protease structures are in the current PDB based on a text search alone. If we do a sequence search, the set of results we get will be much more reliable. 

> Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

We see just one atom per water molecule in this structure because the selected atoms are water, and so the graphical representation of one red sphere represents one water molecule. 

> Q5: There is a conserved water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have (see note below)?

HOH308:O

> Q6: As you have hopefully observed HIV protease is a homodimer (i.e. it is composed of two identical chains). With the aid of the graphic display and the sequence viewer extension can you identify secondary structure elements that are likely to only form in the dimer rather than the monomer?

The secondary structure elements that are likely to only form in the dimer rather than the monomer include the Extended beta conformation, the Alpha helix, the coil, and turn. 

3. Introduction to Bio3D in R

```{r}
library(bio3d)
```

```{r}
pdb <- read.pdb("1hsg")
pdb
```

> Q7: How many amino acid residues are there in this pdb object? 

198

> Q8: Name one of the two non-protein residues? 

HOH

> Q9: How many protein chains are in this structure? 

2

```{r}
attributes(pdb)
```

```{r}
head(pdb$atom)
```

4. Comparative structure analysis of Adenylate Kinase 

> Q10. Which of the packages above is found only on BioConductor and not CRAN?

msa

> Q11. Which of the above packages is not found on BioConductor or CRAN?: 

bio3d-view

> Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket? 

TRUE

```{r}
library(bio3d)
aa <- get.seq("1ake_A")
```

```{r}
aa
```

> Q13. How many amino acids are in this sequence, i.e. how long is this sequence?

214

# Blast or hmmer search 

```{r}
b <- blast.pdb(aa)
```

```{r}
hits <- plot(b)
```
```{r}
head(hits$pdb.id)
```

```{r}
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

```{r}
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```

```{r}
ids <- basename.pdb(pdbs$id)
plot(pdbs, labels=ids)
```
#install.packages("devtools")
```{r}
library(devtools)
```

#install_bitbucket("Grantlab/bio3d-view")
#install.packages("rgl")

```{r}
library(bio3d.view)
```
# install.packages("devtools")
# devtools::install_bitbucket("Grantlab/bio3d/ver_devel/bio3d")

```{r}
library(devtools)
library(bio3d)
```

```{r}
pdb.annotate(c("2mh3_A", "4f3l"), anno.terms = c("structureId", "experimentalTechnique", "resolution","pfam", "source", "citation"))
```

# Perform PCA

```{r}
pc.xray <- pca(pdbs)
plot(pc.xray)
```

# Calculate RMSD
```{r}
rd <- rmsd(pdbs)
```

# Structure-based clustering
```{r}
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)
plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```
# Visualize first principal component
```{r}
pc1 <- mktrj(pc.xray, pc=1, file="pc_1.pdb")
```

```{r}
library(ggplot2)
library(ggrepel)

df <- data.frame(PC1=pc.xray$z[,1], 
                 PC2=pc.xray$z[,2], 
                 col=as.factor(grps.rd),
                 ids=ids)

p <- ggplot(df) + 
  aes(PC1, PC2, col=col, label=ids) +
  geom_point(size=2) +
  geom_text_repel(max.overlaps = 20) +
  theme(legend.position = "none")
p
```

6. Normal mode analysis 

> Q14. What do you note about this plot? Are the black and colored lines similar or different? Where do you think they differ most and why?

The plot shows a significant difference in fluctuation between the black and colored lines, specifically around the residue number 50 and between ~120 and ~155. Upon nucleotide binding, the two conformational states of Adk show different flexibilities around two different binding regions. 
